### PR TITLE
MRTK Instance Application Event Handling / Deterministic Instance Selection

### PR DESCRIPTION
**Pull request by Railboy**
## Overview
Moves all editor / application event listening to static handlers. Previously MixedRealityToolkit instances were responsible for monitoring application state changes individually. In practice this led to dead instances stopping play mode when they couldn't find a config profile.

Registered instances are now sorted by instance ID so that automatic active instance selection is deterministic.

## Changes
- Fixes: #4121

## Verification
- All EditMode tests pass.
- ~~All PlayMode tests pass if executed one at a time. When executed all at once, PressableButtonTests fail due to a GestureRecognizer error. @julenka is this a regression?~~
- All PlayMode tests pass after merging latest changes from dev branch.

### CODE REVIEW

**Comment by wiwei** [comment]
@Railboy regarding the tests, can you check that your fork has the commits from https://github.com/microsoft/MixedRealityToolkit-Unity/pull/4498?

He mentioned that there was some weird physics update bug, though fair warning if the failures you're seeing are completely deterministic in that they happen every single time, it could be a different issue.

**Comment by Railboy** [comment]
@wiwei Tests still fail individually with latest mrtk_development merged in. One-by-one they all pass.

Tried this with the mrtk_development branch and I get the same result, so maybe it's expected behavior?

**Comment by googlan** [comment]
Play mode test error was a bug and should be fixed by #4875 .

File: Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs [hunk 1]
```diff
@@ -848,7 +820,7 @@ private bool RegisterServiceInternal(Type interfaceType, IMixedRealityService se
             if (!isInitializing)
             {
                 serviceInstance.Initialize();
-                serviceInstance.Enable();
+                //serviceInstance.Enable();
```
Comment (Ziugy) [inline Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs]: Is this supposed to be re-enabled or removed?

Comment (Railboy) [reply Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs]: This addresses #4832. I'll remove the comment.

**Comment by julenka** [comment]
@railboy could you also please confirm that you verified #4121 no longer repros, and also that you tested HandInteractionExample scene? I didn't see that in the PR comments but want to make sure that both cases are tested before we complete the merge.

**Comment by Railboy** [comment]
@julenka Correct, with this change in place I can no longer repro #4121 in HandInteractionExample or any other scene.

**Pull request merged**